### PR TITLE
OpenGL backend: implement render passes

### PR DESCRIPTION
**Pull request by kainino0x**
Probably not the most efficient implementation, but works.

Issues:
* [ ] Doesn't seem to render until the second frame. No clue why, yet.
* [ ] Hardcoded 640x480 in more places.
* [ ] Creates new framebuffer objects for every subpass every frame. This should really be done at command buffer build time. (Is that currently possible?)

### CODE REVIEW

File: src/backend/opengl/OpenGLBackend.cpp [hunk 1]
```diff
@@ -100,6 +108,46 @@ namespace opengl {
...
+        glClearColor(0, 0, 0, 1);
+        glStencilMask(0xff);
+        glClearStencil(0);
+        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
+    }
+
+    void Device::InitBackbuffer() {
+        glGenTextures(1, &backTexture);
+        glBindTexture(GL_TEXTURE_2D, backTexture);
+        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB8, 640, 480, 0, GL_RGB, GL_UNSIGNED_BYTE, nullptr);
```
Comment (Kangz) [inline src/backend/opengl/OpenGLBackend.cpp]: The backbuffer could have an alpha channel too :)

File: src/utils/OpenGLBinding.cpp [hunk 2]
```diff
@@ -42,11 +44,18 @@ namespace utils {
...
                 backend::opengl::Init(reinterpret_cast<void*(*)(const char*)>(glfwGetProcAddress), procs, device);
+
+                backendDevice = *device;
+                backend::opengl::InitBackbuffer(backendDevice);
             }
             void SwapBuffers() override {
+                backend::opengl::CommitBackbuffer(backendDevice);
                 glfwSwapBuffers(window);
-                backend::opengl::HACKCLEAR();
+                backend::opengl::HACKCLEAR(backendDevice);
```
Comment (Kangz) [inline src/utils/OpenGLBinding.cpp]: Maybe we should HACKCLEAR just after InitBackbuffer too?

**Comment by Kangz** [comment]
I don't think creating framebuffers all the time is a big problem, they are just container objects in GL. What we could do if we really wanted to reuse framebuffers, would be to create one framebuffer per subpass on nxtFramebuffer creation.

File: src/utils/OpenGLBinding.cpp (hunk 2 above)
Comment (kainino0x) [inline src/utils/OpenGLBinding.cpp]: Ah right, forgot to do this after I moved HACKCLEAR into the device. Unfortunately it doesn't seem to fix the first frame(s) being black.

**Comment by Kangz** [comment]
It depends on how renderpass compatibility is defined, I don't think it should interact a lot with how the framebuffer is set up, it is mostly about load and store ops IIRC.

File: src/backend/opengl/CommandBufferGL.cpp [hunk 3]
```diff
@@ -88,6 +88,10 @@ namespace opengl {
                 case Command::BeginRenderSubpass:
                     {
                         commands.NextCommand<BeginRenderSubpassCmd>();
+                        // TODO(kainino@chromium.org): possible future
+                        // optimization: create these framebuffers at command
```
Comment (Kangz) [inline src/backend/opengl/CommandBufferGL.cpp]: see comment on PR.

Comment (kainino0x) [inline src/backend/opengl/CommandBufferGL.cpp]: ok. I'll edit this comment.

**Comment by kainino0x** [comment]
Skipped CI because I only edited a comment.

**Pull request merged**